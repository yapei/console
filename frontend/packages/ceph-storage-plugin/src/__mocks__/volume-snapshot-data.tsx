export const snapshotData = {
  items: [
    {
      apiVersion: 'snapshot.storage.k8s.io/v1beta1',
      kind: 'VolumeSnapshot',
      metadata: {
        name: 'fakeSnapshot',
        namespace: 'fakes',
      },
      spec: {
        source: {
          persistentVolumeClaimName: 'fake-pvc',
        },
        volumeSnapshotClassName: 'fake-snc',
      },
      status: {
        readyToUse: true,
        restoreSize: '1Gi',
      },
    },
  ],
  breadcrumbs: [
    {
      name: 'Persistent Volume Claims',
      path: `/k8s/ns/fakes/persistentvolumeclaims`,
    },
    {
      name: 'Snapshot Details',
      path: '/k8s/ns/default/kind/example',
    },
  ],
  fakeModel: {
    abbr: 'fk',
    kind: 'fake',
    label: 'Fake',
    labelPlural: 'Fakes',
    plural: 'fakes',
    apiVersion: 'v1',
    apiGroup: 'fakeapi',
  },
};

export const newSnapshotData = {
  items: [
    {
      apiVersion: 'snapshot.storage.k8s.io/v1beta1',
      kind: 'VolumeSnapshot',
      metadata: {
        name: 'fakeSnapshot',
        namespace: 'fakes',
      },
      spec: {
        source: {
          persistentVolumeClaimName: 'fake-pvc',
        },
      },
    },
  ],
};

export const pvcData = {
  data: {
    apiVersion: 'v1',
    kind: 'PersistentVolumeClaim',
    metadata: {
      name: 'fake-pvc',
      namespace: 'fakes',
    },
    spec: {
      resources: {
        requests: {
          storage: '1Gi',
        },
      },
      storageClassName: 'fake-scn',
    },
    status: {
      phase: 'Bound',
      accessModes: ['ReadWriteOnce'],
      capacity: {
        storage: '1Gi',
      },
    },
  },
};
