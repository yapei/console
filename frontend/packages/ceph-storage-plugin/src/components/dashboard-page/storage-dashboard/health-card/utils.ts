import * as _ from 'lodash';
import { HealthState } from '@console/shared/src/components/dashboard/health-card/states';
import { PrometheusHealthHandler } from '@console/plugin-sdk';

const CephHealthStatus = [
  {
    state: HealthState.OK,
  },
  {
    state: HealthState.WARNING,
  },
  {
    state: HealthState.ERROR,
  },
  {
    state: HealthState.UNKNOWN,
  },
];

export const getCephHealthState: PrometheusHealthHandler = (responses = [], errors = []) => {
  if (errors.length) {
    return CephHealthStatus[3];
  }
  if (!responses[0]) {
    return { state: HealthState.LOADING };
  }

  const value = _.get(responses[0], 'data.result[0].value[1]');
  return CephHealthStatus[value] || CephHealthStatus[3];
};
