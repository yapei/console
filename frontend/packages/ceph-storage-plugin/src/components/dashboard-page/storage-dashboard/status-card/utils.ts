import { PrometheusHealthHandler, ResourceHealthHandler } from '@console/plugin-sdk';
import { HealthState } from '@console/shared/src/components/dashboard/status-card/states';
import { getResiliencyProgress } from '../../../../utils';
import { WatchCephResource } from '../../../../types';

const CephHealthStatus = {
  HEALTH_OK: {
    state: HealthState.OK,
  },
  HEALTH_WARN: {
    state: HealthState.WARNING,
  },
  HEALTH_ERR: {
    state: HealthState.ERROR,
    message: 'Error',
  },
};

export const getCephHealthState: ResourceHealthHandler<WatchCephResource> = ({ ceph }) => {
  const { data, loaded, loadError } = ceph;
  const status = data?.[0]?.status?.ceph?.health;

  if (loadError) {
    return { state: HealthState.NOT_AVAILABLE };
  }
  if (!loaded) {
    return { state: HealthState.LOADING };
  }
  if (data.length === 0) {
    return { state: HealthState.NOT_AVAILABLE };
  }
  return CephHealthStatus[status] || { state: HealthState.UNKNOWN };
};

export const getDataResiliencyState: PrometheusHealthHandler = (responses) => {
  const progress: number = getResiliencyProgress(responses[0].response);
  if (responses[0].error) {
    return { state: HealthState.NOT_AVAILABLE };
  }
  if (!responses[0].response) {
    return { state: HealthState.LOADING };
  }
  if (Number.isNaN(progress)) {
    return { state: HealthState.UNKNOWN };
  }
  if (progress < 1) {
    return { state: HealthState.PROGRESS, message: 'Progressing' };
  }
  return { state: HealthState.OK };
};
