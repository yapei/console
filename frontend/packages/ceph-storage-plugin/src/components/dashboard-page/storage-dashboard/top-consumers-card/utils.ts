import * as _ from 'lodash';
import { humanizeBinaryBytesWithoutB } from '@console/internal/components/utils';
import { PrometheusResponse, DataPoint } from '@console/internal/components/graphs';

export const getMetricType = (resource, metricType) => _.get(resource, ['metric', metricType], '');

export const getGraphVectorStats: GetStats = (response, metricType, unit) => {
  return response.map((r) => {
    return r.values.map((arr) => ({
      name: getMetricType(r, metricType),
      x: new Date(arr[0] * 1000),
      y: Number(humanizeBinaryBytesWithoutB(arr[1], null, unit).value),
    }));
  });
};

export const sortResources: SortResourcesProps = (a, b) => {
  const aVal = _.get(a, 'values');
  const bVal = _.get(b, 'values');
  const x = _.get(a, ['values', aVal.length - 1, 1]);
  const y = _.get(b, ['values', bVal.length - 1, 1]);
  return y - x;
};

type GetStats = (
  response: PrometheusResponse['data']['result'],
  metric?: string,
  unit?: string,
) => DataPoint[][];

type SortResourcesProps = (
  a: PrometheusResponse['data']['result'],
  b: PrometheusResponse['data']['result'],
) => number;
