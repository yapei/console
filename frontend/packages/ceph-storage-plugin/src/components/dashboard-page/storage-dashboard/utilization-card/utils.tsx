import * as _ from 'lodash';
import { DataPoint } from '@console/internal/components/graphs';
import { Humanize, humanizeNumber } from '@console/internal/components/utils';

export const humanizeIOPS: Humanize = (value) => {
  const humanizedNumber = humanizeNumber(value);
  const unit = 'IOPS';
  return {
    ...humanizedNumber,
    string: `${humanizedNumber.value} ${unit}`,
    unit,
  };
};

export const humanizeLatency: Humanize = (value) => {
  const humanizedNumber = humanizeNumber(value);
  const unit = 'ms';
  return {
    ...humanizedNumber,
    string: `${humanizedNumber.value} ${unit}`,
    unit,
  };
};

export const getLatestValue = (stats: DataPoint[]) => Number(_.get(stats[stats.length - 1], 'y'));
