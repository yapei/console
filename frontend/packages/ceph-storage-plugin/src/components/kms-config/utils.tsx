import * as React from 'react';
import * as _ from 'lodash';
import { K8sResourceKind, ConfigMapKind, SecretKind } from '@console/internal/module/k8s/types';
import { k8sCreate, k8sPatch } from '@console/internal/module/k8s/resource';
import { ConfigMapModel, SecretModel } from '@console/internal/models';
import {
  CEPH_STORAGE_NAMESPACE,
  MODES,
  KMSConfigMapName,
  KMSSecretName,
  KMSConfigMapCSIName,
} from '../../constants';
import { Action } from '../ocs-install/attached-devices-mode/reducer';
import { InternalClusterAction } from '../ocs-install/internal-mode/reducer';
import { KMSConfig, KMSConfigMap } from '../../types';
import { StorageClassClusterAction } from '../../utils/kms-encryption';

export const parseURL = (url: string) => {
  try {
    return new URL(url);
  } catch (e) {
    return null;
  }
};

export const generateCASecret = (caCertificate: string) => ({
  apiVersion: SecretModel.apiVersion,
  kind: SecretModel.kind,
  metadata: {
    name: `ocs-kms-ca-secret-${Math.random()
      .toString(36)
      .substring(7)}`,
    namespace: CEPH_STORAGE_NAMESPACE,
  },
  stringData: {
    cert: caCertificate,
  },
});

export const generateClientSecret = (clientCertificate: string) => ({
  apiVersion: SecretModel.apiVersion,
  kind: SecretModel.kind,
  metadata: {
    name: `ocs-kms-client-cert-${Math.random()
      .toString(36)
      .substring(7)}`,
    namespace: CEPH_STORAGE_NAMESPACE,
  },
  stringData: {
    cert: clientCertificate,
  },
});

export const generateClientKeySecret = (clientKey: string) => ({
  apiVersion: SecretModel.apiVersion,
  kind: SecretModel.kind,
  metadata: {
    name: `ocs-kms-client-key-${Math.random()
      .toString(36)
      .substring(7)}`,
    namespace: CEPH_STORAGE_NAMESPACE,
  },
  stringData: {
    key: clientKey,
  },
});

export const createKmsResources = (kms: KMSConfig, update = false, previousData?: any) => {
  let tokenSecret: SecretKind;
  if (kms.token) {
    tokenSecret = {
      apiVersion: SecretModel.apiVersion,
      kind: SecretModel.kind,
      metadata: {
        name: KMSSecretName,
        namespace: CEPH_STORAGE_NAMESPACE,
      },
      stringData: {
        token: kms.token.value,
      },
    };
  }

  const resources: Promise<K8sResourceKind>[] = [];

  const parsedAddress = parseURL(kms.address.value);

  const configData: KMSConfigMap = {
    KMS_PROVIDER: 'vault',
    KMS_SERVICE_NAME: kms.name.value,
    VAULT_ADDR: `${`${parsedAddress.protocol}//${parsedAddress.hostname}`}:${kms.port.value}`,
    VAULT_BACKEND_PATH: kms.backend,
    VAULT_CACERT: kms.caCert?.metadata.name,
    VAULT_TLS_SERVER_NAME: kms.tls,
    VAULT_CLIENT_CERT: kms.clientCert?.metadata.name,
    VAULT_CLIENT_KEY: kms.clientKey?.metadata.name,
    VAULT_NAMESPACE: kms.providerNamespace,
  };

  const configMapObj: ConfigMapKind = {
    apiVersion: ConfigMapModel.apiVersion,
    kind: ConfigMapModel.kind,
    data: {
      ...configData,
    },
    metadata: {
      name: KMSConfigMapName,
      namespace: CEPH_STORAGE_NAMESPACE,
    },
  };

  const csiConfigData: KMSConfigMap = {
    KMS_PROVIDER: 'vaulttokens',
    KMS_SERVICE_NAME: kms.name.value,
    VAULT_ADDR: `${`${parsedAddress.protocol}//${parsedAddress.hostname}`}:${kms.port.value}`,
    VAULT_BACKEND_PATH: kms.backend,
    VAULT_CACERT: kms.caCert?.metadata.name,
    VAULT_TLS_SERVER_NAME: kms.tls,
    VAULT_CLIENT_CERT: kms.clientCert?.metadata.name,
    VAULT_CLIENT_KEY: kms.clientKey?.metadata.name,
    VAULT_NAMESPACE: kms.providerNamespace,
    VAULT_TOKEN_NAME: KMSSecretName,
    VAULT_CACERT_FILE: kms.caCertFile,
    VAULT_CLIENT_CERT_FILE: kms.clientCertFile,
    VAULT_CLIENT_KEY_FILE: kms.clientKeyFile,
  };

  const csiConfigObj: ConfigMapKind = {
    apiVersion: ConfigMapModel.apiVersion,
    kind: ConfigMapModel.kind,
    data: {
      [`1-${kms.name.value}`]: JSON.stringify(csiConfigData),
    },
    metadata: {
      name: KMSConfigMapCSIName,
      namespace: CEPH_STORAGE_NAMESPACE,
    },
  };

  if (kms.caCert) {
    resources.push(k8sCreate(SecretModel, kms.caCert, { ns: CEPH_STORAGE_NAMESPACE }));
  }

  if (kms.clientCert) {
    resources.push(k8sCreate(SecretModel, kms.clientCert, { ns: CEPH_STORAGE_NAMESPACE }));
  }

  if (kms.clientKey) {
    resources.push(k8sCreate(SecretModel, kms.clientKey, { ns: CEPH_STORAGE_NAMESPACE }));
  }

  if (update) {
    const patchValue = Object.keys(previousData).length + 1;
    const cmPatch = [
      {
        op: 'replace',
        path: `/data/${patchValue}-${kms.name.value}`,
        value: JSON.stringify(csiConfigData),
      },
    ];
    resources.push(k8sPatch(ConfigMapModel, csiConfigObj, cmPatch));
  } else {
    resources.push(k8sCreate(SecretModel, tokenSecret, { ns: CEPH_STORAGE_NAMESPACE }));
    resources.push(k8sCreate(ConfigMapModel, configMapObj, { ns: CEPH_STORAGE_NAMESPACE }));
    resources.push(k8sCreate(ConfigMapModel, csiConfigObj, { ns: CEPH_STORAGE_NAMESPACE }));
  }

  return resources;
};

export const setEncryptionDispatch = (
  keyType: any,
  mode: string,
  dispatch: React.Dispatch<Action | InternalClusterAction | StorageClassClusterAction>,
  valueType?: any,
) => {
  const stateType = mode === MODES.ATTACHED_DEVICES ? _.camelCase(keyType) : keyType;

  if (valueType) {
    const stateValue =
      mode === MODES.ATTACHED_DEVICES ? { value: valueType } : { payload: valueType };
    dispatch({ type: stateType, ...stateValue });
  } else {
    dispatch({ type: stateType });
  }
};

export const getPort = (url: URL) => {
  if (url.port === '') {
    return url.protocol === 'http:' ? '80' : '443';
  }
  return url.port;
};

export const scKmsConfigValidation = (kms: KMSConfig): boolean =>
  kms.name?.valid &&
  kms.address?.valid &&
  kms.port?.valid &&
  kms.name.value !== '' &&
  kms.address.value !== '' &&
  kms.port.value !== '';

export const kmsConfigValidation = (kms: KMSConfig): boolean =>
  kms.token?.valid && kms.token.value !== '' && scKmsConfigValidation(kms);
