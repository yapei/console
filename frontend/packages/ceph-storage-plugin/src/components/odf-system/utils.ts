import { HumanizeResult } from '@console/internal/components/utils';
import { PrometheusResponse } from '@console/internal/module/k8s';

// Operator uses`<kind>.<apiGroup>/<apiVersion>`
export const getGVK = (label: string) => {
  const kind = label.slice(0, label.indexOf('.'));
  const apiGroup = label.slice(label.indexOf('.') + 1, label.indexOf('/'));
  const apiVersion = label.slice(label.indexOf('/') + 1, label.length);
  return { kind, apiGroup, apiVersion };
};

export type SystemMetrics = {
  metrics: {
    [systeName: string]: {
      rawCapacity: HumanizeResult;
      usedCapacity: HumanizeResult;
      iops: HumanizeResult;
      throughput: HumanizeResult;
      latency: HumanizeResult;
    };
  };
};

type MetricNormalize = (
  latency: PrometheusResponse,
  throughput: PrometheusResponse,
  rawCapacity: PrometheusResponse,
  usedCapacity: PrometheusResponse,
  iops: PrometheusResponse,
) => SystemMetrics['metrics'];

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const normalizeMetrics: MetricNormalize = (
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _latency,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _throughput,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _rawCapacity,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _usedCapacity,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _iops,
) => {
  // Todo(bipuladh): Add parsing logic for above items
  return {
    metrics: {} as any,
  };
};
