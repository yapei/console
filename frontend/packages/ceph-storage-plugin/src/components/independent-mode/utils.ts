import * as _ from 'lodash';
import { ClusterServiceVersionKind } from '@console/operator-lifecycle-manager';
import { HealthState } from '@console/shared/src/components/dashboard/status-card/states';
import { DataValidator, DataState, ErrorType, Field } from './types';
import { K8sResourceKind } from '@console/internal/module/k8s';

export const getValidJSON: DataValidator = (fData) => {
  try {
    // Just see if it fails
    const parsedData = JSON.parse(fData);
    return { isValid: true, parsedData };
  } catch (e) {
    return { isValid: false, errorMessage: 'File is not a valid JSON file.' };
  }
};

export const checkError = (data: DataState): ErrorType[] => {
  const errors: ErrorType[] = [];
  for (const key in data) {
    if (!_.isNull(data[key]) && _.isEmpty(_.trim(data[key])))
      errors.push({ field: key as Field, message: `${key} cannot be empty` });
    else errors.push({ field: key as Field, message: '' });
  }
  return errors;
};

export const checkForIndependentSupport = (csv: ClusterServiceVersionKind): boolean => {
  const independent: string =
    csv.metadata.annotations?.['external.cluster.ocs.openshift.io/supported'];
  return independent === 'true';
};

enum ClusterPhase {
  CONNECTED = 'Connected',
  READY = 'Ready',
  CONNECTING = 'Connecting',
  PROGRESSING = 'Progressing',
  ERROR = 'Error',
}

const PhaseToState = Object.freeze({
  [ClusterPhase.CONNECTED]: HealthState.OK,
  [ClusterPhase.READY]: HealthState.OK,
  [ClusterPhase.CONNECTING]: HealthState.UPDATING,
  [ClusterPhase.PROGRESSING]: HealthState.UPDATING,
  [ClusterPhase.ERROR]: HealthState.ERROR,
});

export const getClusterHealth = (cluster: K8sResourceKind, loaded: boolean, error): HealthState => {
  const phase = cluster?.status?.phase;
  if (!_.isEmpty(error)) {
    if (error?.response?.status === 404) return HealthState.NOT_AVAILABLE;
    return HealthState.ERROR;
  }
  if (!loaded) return HealthState.LOADING;
  if (!_.isEmpty(cluster)) return PhaseToState[phase];
  return HealthState.NOT_AVAILABLE;
};
