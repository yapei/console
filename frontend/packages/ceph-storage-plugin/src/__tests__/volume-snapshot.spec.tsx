import * as React from 'react';
import { ShallowWrapper, shallow } from 'enzyme';

import {
  DetailsPage,
  ListPage,
  RowFunctionArgs,
  TableProps,
  Table,
} from '@console/internal/components/factory';
import { K8sResourceKind } from '@console/internal/module/k8s';
import {
  ResourceLink,
  ResourceKebab,
  PageComponentProps,
} from '@console/internal/components/utils';
import { Status } from '@console/shared';

import {
  VolumeSnapshotDetails,
  VolumeSnapshotPage,
  VolumeSnapshotList,
  VolumeSnapshotTableHeader,
  VolumeSnapshotTableRow,
  DetailsComponent,
} from '../components/volume-snapshot/volume-snapshot';

import { DeleteSnapshot } from '../utils/delete-snapshot-workflow';
import { RestorePVC } from '../utils/restore-pvc-workflow';
import { snapshotData, pvcData } from '../__mocks__/volume-snapshot-data';

describe('<VolumeSnapshotList>', () => {
  let wrapper: ShallowWrapper<TableProps>;
  let rowArgs: RowFunctionArgs<K8sResourceKind>;

  beforeEach(() => {
    wrapper = shallow(
      <VolumeSnapshotList
        aria-label="Volume Snapshot"
        data={snapshotData.items}
        Header={VolumeSnapshotTableHeader}
        Row={VolumeSnapshotTableRow}
      />,
    );
    rowArgs = {
      obj: snapshotData.items[0],
      index: 1,
      key: '1',
      style: {},
    } as any;
  });

  it('renders `Table` with correct props', () => {
    expect(wrapper.find(Table).exists()).toBe(true);
  });

  it('should render the proper Headers in the Resources tab', () => {
    const expectedHeader: string[] = ['Name', 'Date', 'Status', 'Size'];

    const headers = wrapper
      .find(Table)
      .props()
      .Header();

    expectedHeader.forEach((header, i) => {
      expect(headers[i].title).toBe(header);
    });
  });

  it('should render the TableRow component', () => {
    const volumeSnapshotTableRow = shallow(VolumeSnapshotTableRow(rowArgs));
    expect(volumeSnapshotTableRow.find('tr').exists()).toBe(true);
  });

  it('should render the number of snapshots deployed for resources that support it', () => {
    const volumeSnapshotTableRow = shallow(VolumeSnapshotTableRow(rowArgs));
    expect(volumeSnapshotTableRow.find(Status).exists()).toBe(true);
    expect(volumeSnapshotTableRow.find(Status).props().status).toBe('Ready');

    expect(volumeSnapshotTableRow.find(ResourceLink).exists()).toBe(true);
    expect(volumeSnapshotTableRow.find(ResourceLink).props().title).toEqual('fakeSnapshot');

    expect(volumeSnapshotTableRow.find(ResourceKebab).exists()).toBe(true);
    expect(volumeSnapshotTableRow.find(ResourceKebab).props().actions).toEqual([
      RestorePVC,
      DeleteSnapshot,
    ]);
  });
});

describe('<VolumeSnapshotPage>', () => {
  let wrapper: ShallowWrapper<PageComponentProps>;

  beforeEach(() => {
    wrapper = shallow(
      <VolumeSnapshotPage
        match={{
          url: '/k8s/ns/default/kind/example',
          path: '/k8s/ns/:ns/:plural/:name',
          isExact: true,
          params: {},
        }}
        obj={pvcData.data}
      />,
    );
  });

  it('renders `ListPage` with correct props', () => {
    expect(wrapper.find(ListPage).exists()).toBe(true);
  });
});

describe('<VolumeSnapshotDetails>', () => {
  let wrapper: ShallowWrapper<any>;
  const detailArgs = {
    obj: snapshotData.items[0],
  } as any;

  beforeEach(() => {
    wrapper = shallow(
      <VolumeSnapshotDetails
        match={{
          url: '/k8s/ns/default/kind/example',
          path: '/k8s/ns/:ns/:plural/:name',
          isExact: true,
          params: {
            ns: 'fakes',
          },
        }}
        kind="VolumeSnapshotModel"
      />,
    );
  });

  it('renders `DetailsPage` with correct props', () => {
    const detailsWrapper = wrapper.find(DetailsPage);
    expect(detailsWrapper.exists()).toBe(true);
  });

  it('should render `DetailsPage` Component with proper page tabs', () => {
    expect(
      wrapper
        .find(DetailsPage)
        .props()
        .pages.map((p) => p.name),
    ).toEqual(['Details']);
  });

  it('should render the Details component', () => {
    const detailsWrapper = wrapper.find(DetailsPage);
    expect(detailsWrapper.props().pages[0].component).toEqual(DetailsComponent);
    expect(DetailsComponent(detailArgs)).toMatchSnapshot();
  });

  it('should render the details action menu', () => {
    const detailsWrapper = wrapper.find(DetailsPage);
    expect(detailsWrapper.props().menuActions).toEqual([RestorePVC, DeleteSnapshot]);
  });

  it('should render the breadcrumbs', () => {
    const detailsWrapper = wrapper.find(DetailsPage);
    expect(detailsWrapper.props().breadcrumbsFor(snapshotData.items[0])).toEqual(
      snapshotData.breadcrumbs,
    );
  });
});
