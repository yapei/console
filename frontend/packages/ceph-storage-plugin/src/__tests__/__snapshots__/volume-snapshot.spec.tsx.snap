// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VolumeSnapshotDetails> should render the Details component 1`] = `
<UNDEFINED>
  <div
    className="co-m-pane__body"
  >
    <SectionHeading
      text="Volume Snapshot Details"
    />
    <div
      className="row"
    >
      <div
        className="col-md-6"
      >
        <ResourceSummary
          resource={
            Object {
              "apiVersion": "snapshot.storage.k8s.io/v1beta1",
              "kind": "VolumeSnapshot",
              "metadata": Object {
                "name": "fakeSnapshot",
                "namespace": "fakes",
              },
              "spec": Object {
                "source": Object {
                  "persistentVolumeClaimName": "fake-pvc",
                },
                "volumeSnapshotClassName": "fake-snc",
              },
              "status": Object {
                "readyToUse": true,
                "restoreSize": "1Gi",
              },
            }
          }
        >
          <dt>
            Status
          </dt>
          <dd>
            Ready
          </dd>
          <dt>
            Size
          </dt>
          <dd>
            1Gi
          </dd>
        </ResourceSummary>
      </div>
      <div
        className="col-md-6"
      >
        <dt>
          Persistent Volume Claim
        </dt>
        <dd>
          <UNDEFINED
            kind="PersistentVolumeClaim"
            name="fake-pvc"
            namespace="fakes"
          />
        </dd>
      </div>
    </div>
  </div>
</UNDEFINED>
`;
